virtio-devices/src/vhost_user/vhost_user_device.rs

```diff
@@ -0,0 +1,426 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::io::Write as _;
+use std::sync::atomic::AtomicBool;
+use std::sync::{Arc, Barrier, Mutex};
+use std::{result, thread};
+
+use seccompiler::SeccompAction;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use vhost::vhost_user::message::{
+    VhostUserConfigFlags, VhostUserProtocolFeatures, VhostUserVirtioFeatures,
+};
+use vhost::vhost_user::{FrontendReqHandler, VhostUserFrontend, VhostUserFrontendReqHandler};
+use virtio_queue::Queue;
+use vm_device::UserspaceMapping;
+use vm_memory::{ByteValued, GuestMemoryAtomic};
+use vm_migration::protocol::MemoryRangeTable;
+use vm_migration::{Migratable, MigratableError, Pausable, Snapshot, Snapshottable, Transportable};
+use vmm_sys_util::eventfd::EventFd;
+
+use super::vu_common_ctrl::VhostUserHandle;
+use super::{Error, Result};
+use crate::seccomp_filters::Thread;
+use crate::thread_helper::spawn_virtio_thread;
+use crate::vhost_user::VhostUserCommon;
+use crate::{
+    ActivateResult, GuestMemoryMmap, GuestRegionMmap, MmapRegion, VIRTIO_F_IOMMU_PLATFORM,
+    VirtioCommon, VirtioDevice, VirtioInterrupt, VirtioSharedMemoryList,
+};
+
+const NUM_QUEUE_OFFSET: usize = 1;
+const DEFAULT_QUEUE_NUMBER: usize = 2;
+
+#[derive(Serialize, Deserialize)]
+pub struct State {
+    pub avail_features: u64,
+    pub acked_features: u64,
+    pub config: VirtioVhostUserConfig,
+    pub acked_protocol_features: u64,
+    pub vu_num_queues: usize,
+    pub backend_req_support: bool,
+}
+
+struct BackendReqHandler {}
+impl VhostUserFrontendReqHandler for BackendReqHandler {}
+
+pub const VIRTIO_FS_TAG_LEN: usize = 36;
+#[serde_as]
+#[derive(Copy, Clone, Serialize, Deserialize, Default)]
+#[repr(C, packed)]
+pub struct VirtioVhostUserConfig {
+    pub num_request_queues: u32,
+}
```

Comment:
Would it be better to just store `num_request_queues` directly?  Is it likely other fields will be added?  "Config" sounds like it would be the virtio device config space.

Reply:
Still relevant.